### PR TITLE
Add TransferAllTokensService and tests.

### PR DESCRIPTION
**Pull request by mikebarile**

### CODE REVIEW

File: src/main/services/token/TransferAllTokensService.kt [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+object TransferAllTokensService {
+    fun execute(
+            caller: UserAccount,
+            from: String,
+            to: String,
+            notes: String?) : SOAResult<TransactionList> {
+        val address = caller.cryptoKeyPair.publicKey
+
+        // verify that the caller is the from address
+        if(address != from)
```
Comment (lostcodingsomewhere) [inline src/main/services/token/TransferAllTokensService.kt]: could just have no from -- and just use the callers address

File: src/main/services/token/TransferAllTokensService.kt [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        // Transfer all tokens to the new address
+        var resultingTransactions = mutableListOf<Transaction>()
+        var result : SOAResult<TransactionList> =
+                SOAResult(SOAResultType.SUCCESS, "All tokens transferred successfully.", null)
+
+        if(mapOfBalances.size == 0)
+            result.message = "Address has no associated balances so no tokens were transferred."
+
+        mapOfBalances.forEach { tokenId, balance ->
+            if(balance <= 0.0 || result.result != SOAResultType.SUCCESS)
```
Comment (lostcodingsomewhere) [inline src/main/services/token/TransferAllTokensService.kt]: if the balance is less than 0 should we change result to failure and add message?

File: src/main/services/token/TransferTokenService.kt [hunk 3]
```diff
@@ -35,4 +35,30 @@ object TransferTokenService {
             notes
         )
     }
+
+    fun execute(
```
Comment (lostcodingsomewhere) [inline src/main/services/token/TransferTokenService.kt]: why do we need another execute?

Comment (mikebarile) [reply src/main/services/token/TransferTokenService.kt]: has a different signature

Comment (lostcodingsomewhere) [reply src/main/services/token/TransferTokenService.kt]: could you just add token id to the other one?

Comment (mikebarile) [reply src/main/services/token/TransferTokenService.kt]: right, optional parameters :D :D :D

Comment (lostcodingsomewhere) [reply src/main/services/token/TransferTokenService.kt]: FTW

Comment (lostcodingsomewhere) [reply src/main/services/token/TransferTokenService.kt]: and extra nice is optional params where u set a default value and put those at the end of the method signature...makes calling the methods super clean!

File: src/main/services/token/TransferTokenService.kt [hunk 4]
```diff
@@ -14,8 +14,9 @@ object TransferTokenService {
         caller: UserAccount,
         from: String,
         to: String,
-        name: String,
         amount: Double,
+        name: String?,
```
Comment (lostcodingsomewhere) [inline src/main/services/token/TransferTokenService.kt]: maybe nice to default these to null so you dont have to pass null, null

Comment (lostcodingsomewhere) [reply src/main/services/token/TransferTokenService.kt]: null, null, null, null lol

File: src/main/services/token/TransferAllTokensService.kt [hunk 5]
```diff
@@ -13,24 +13,16 @@ import main.services.reward.DistributeRewardService
 object TransferAllTokensService {
     fun execute(
             caller: UserAccount,
-            from: String,
-            to: String,
+            toAddress: String,
             notes: String?) : SOAResult<TransactionList> {
```
Comment (lostcodingsomewhere) [inline src/main/services/token/TransferAllTokensService.kt]: default notes to null

**Comment by lostcodingsomewhere** [comment]
basically suggest moving lesser used fields to the end of a function call and defaulting them to null -- makes calling a lot cleaner -- general rule of thumb

**Comment by mikebarile** [comment]
@lostcodingsomewhere back at ya

File: src/main/services/token/TransferAllTokensService.kt [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+ * Transfer all tokens held by one address to another
+ */
+object TransferAllTokensService {
+    fun execute(
+            caller: UserAccount,
+            toAddress: String,
+            notes: String? = null) : SOAResult<TransactionList> {
+        val fromAddress = caller.cryptoKeyPair.publicKey
+
+        // Generate map of all tokenIds with a balance
```
Comment (lostcodingsomewhere) [inline src/main/services/token/TransferAllTokensService.kt]: maybe this mapOfBalances thing should be moved to a helper function

File: src/main/services/token/TransferAllTokensService.kt [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        val transactionsResult = TransferTokenHelper.getTransferHistory(fromAddress, null)
```
Comment (lostcodingsomewhere) [inline src/main/services/token/TransferAllTokensService.kt]: maybe refactor getTransferHistory second param to default to null

File: src/main/services/token/TransferTokenService.kt [hunk 8]
```diff
@@ -14,25 +14,39 @@ object TransferTokenService {
...
-            from,
-            to,
-            name,
-            amount,
-            ActionType.TRANSFER,
-            previousTransactionId,
-            notes
-        )
+        if (tokenId != null)
+            return TransferTokenHelper.transferToken(
```
Comment (lostcodingsomewhere) [inline src/main/services/token/TransferTokenService.kt]: maybe we should refactor transferToken to handle this -- basically will have both optional params -- and error if both are null -- and will have a private method that handels getting the token

**Comment by mikebarile** [comment]
@lostcodingsomewhere back at ya

**Pull request merged**